File: apps/web/src/lib/rbac/role-permissions.ts
import { UserRole } from "@/types/api";
import { Permission, type PermissionKey } from "./permissions";

/**
 * Maps each role to its set of permissions.
 * ADMIN has all permissions.
 * ASSISTANT_MANAGER has most admin permissions except cost visibility and user management.
 * EMPLOYEE has limited permissions.
 */
export const ROLE_PERMISSIONS: Record<UserRole, ReadonlySet<PermissionKey>> = {
  [UserRole.ADMIN]: new Set(Object.values(Permission)),

  [UserRole.ASSISTANT_MANAGER]: new Set([
    // Dashboard
    Permission.DASHBOARD_VIEW,

    // Products - full CRUD (but no cost visibility)
    Permission.PRODUCTS_VIEW,
    Permission.PRODUCTS_CREATE,
    Permission.PRODUCTS_UPDATE,
    Permission.PRODUCTS_DELETE,

    // Storage - full access
    Permission.STORAGE_VIEW,
    Permission.STORAGE_CREATE,
    Permission.STORAGE_UPDATE,
    Permission.STORAGE_DELETE,

    // Inventory operations - full access
    Permission.INVENTORY_ADJUST,
    Permission.INVENTORY_TRANSFER,

    // Shipments - full CRUD (but no cost visibility)
    Permission.SHIPMENTS_VIEW,
    Permission.SHIPMENTS_CREATE,
    Permission.SHIPMENTS_UPDATE,
    Permission.SHIPMENTS_DELETE,
    Permission.SHIPMENTS_RECEIVE,

    // Analytics
    Permission.ANALYTICS_VIEW,

    // Notifications - full access
    Permission.NOTIFICATIONS_VIEW,
    Permission.NOTIFICATIONS_MANAGE,

    // Audit Log
    Permission.AUDIT_LOG_VIEW,

    // Team - view only (no user management)
    Permission.TEAM_VIEW,

    // Settings - full access
    Permission.SETTINGS_VIEW,
    Permission.SETTINGS_MANAGE,

    // Reviews - full access
    Permission.REVIEWS_VIEW,
    Permission.REVIEWS_MANAGE,

    // Machine Displays - full access
    Permission.MACHINE_DISPLAYS_VIEW,
    Permission.MACHINE_DISPLAYS_MANAGE,

    // EXCLUDED (Admin-only):
    // - Permission.COSTS_VIEW
    // - Permission.USERS_MANAGE
  ]),

  [UserRole.EMPLOYEE]: new Set([
    // Products - view only
    Permission.PRODUCTS_VIEW,

    // Storage - view only
    Permission.STORAGE_VIEW,

    // Inventory operations - full access
    Permission.INVENTORY_ADJUST,
    Permission.INVENTORY_TRANSFER,

    // Shipments - view and receive only (no create, update, or delete)
    Permission.SHIPMENTS_VIEW,
    Permission.SHIPMENTS_RECEIVE,

    // Audit Log - view only
    Permission.AUDIT_LOG_VIEW,

    // Reviews - view only
    Permission.REVIEWS_VIEW,

    // Settings - view only
    Permission.SETTINGS_VIEW,

    // Notifications - view only
    Permission.NOTIFICATIONS_VIEW,

    // Machine Displays - full access for employees to manage displays
    Permission.MACHINE_DISPLAYS_VIEW,
    Permission.MACHINE_DISPLAYS_MANAGE,
  ]),
};

/**
 * Check if a role has a specific permission.
 */
export function hasPermission(
  role: UserRole | string | undefined,
  permission: PermissionKey
): boolean {
  if (!role) return false;
  // Normalize role to uppercase to match UserRole enum keys
  const normalizedRole = (typeof role === "string" ? role.toUpperCase() : role) as UserRole;
  const permissions = ROLE_PERMISSIONS[normalizedRole];
  return permissions?.has(permission) ?? false;
}

/**
 * Check if a role has all of the specified permissions.
 */
export function hasAllPermissions(
  role: UserRole | undefined,
  permissions: readonly PermissionKey[]
): boolean {
  if (!role) return false;
  return permissions.every((p) => hasPermission(role, p));
}

/**
 * Check if a role has any of the specified permissions.
 */
export function hasAnyPermission(
  role: UserRole | undefined,
  permissions: readonly PermissionKey[]
): boolean {
  if (!role) return false;
  return permissions.some((p) => hasPermission(role, p));
}

File: services/inventory-service/src/main/java/com/mirai/inventoryservice/identity/domain/RolePermissions.java
package com.mirai.inventoryservice.identity.domain;

import org.springframework.security.core.Authentication;

import java.util.EnumMap;
import java.util.EnumSet;
import java.util.Map;
import java.util.Set;

import static com.mirai.inventoryservice.identity.domain.Permission.*;

/**
 * Backend-owned role-to-permission matrix. Mirrors
 * {@code apps/web/src/lib/rbac/role-permissions.ts}: ADMIN has every permission,
 * ASSISTANT_MANAGER has everything except cost visibility and user management,
 * EMPLOYEE has a limited operational subset. See RolePermissionsTest for the
 * cross-side parity assertions ported from role-permissions.test.ts.
 */
public final class RolePermissions {

    private static final Map<UserRole, Set<Permission>> BY_ROLE = new EnumMap<>(UserRole.class);

    static {
        BY_ROLE.put(UserRole.ADMIN, EnumSet.allOf(Permission.class));

        BY_ROLE.put(UserRole.ASSISTANT_MANAGER, EnumSet.of(
                DASHBOARD_VIEW,

                PRODUCTS_VIEW, PRODUCTS_CREATE, PRODUCTS_UPDATE, PRODUCTS_DELETE,

                STORAGE_VIEW, STORAGE_CREATE, STORAGE_UPDATE, STORAGE_DELETE,

                INVENTORY_ADJUST, INVENTORY_TRANSFER,

                SHIPMENTS_VIEW, SHIPMENTS_CREATE, SHIPMENTS_UPDATE, SHIPMENTS_DELETE, SHIPMENTS_RECEIVE,

                ANALYTICS_VIEW,

                NOTIFICATIONS_VIEW, NOTIFICATIONS_MANAGE,

                AUDIT_LOG_VIEW,

                TEAM_VIEW,

                SETTINGS_VIEW, SETTINGS_MANAGE,

                REVIEWS_VIEW, REVIEWS_MANAGE,

                MACHINE_DISPLAYS_VIEW, MACHINE_DISPLAYS_MANAGE,

                MSRP_VIEW, KUJI_PRICES_VIEW
                // Excluded (admin-only): COSTS_VIEW, USERS_MANAGE
        ));

        BY_ROLE.put(UserRole.EMPLOYEE, EnumSet.of(
                PRODUCTS_VIEW,

                STORAGE_VIEW,

                INVENTORY_ADJUST, INVENTORY_TRANSFER,

                SHIPMENTS_VIEW, SHIPMENTS_RECEIVE,

                AUDIT_LOG_VIEW,

                REVIEWS_VIEW,

                SETTINGS_VIEW,

                NOTIFICATIONS_VIEW,

                MACHINE_DISPLAYS_VIEW, MACHINE_DISPLAYS_MANAGE,

                TEAM_VIEW
        ));
    }

    private RolePermissions() {
    }

    public static Set<Permission> forRole(UserRole role) {
        if (role == null) {
            return Set.of();
        }
        return BY_ROLE.getOrDefault(role, Set.of());
    }

    public static boolean hasPermission(UserRole role, Permission permission) {
        return forRole(role).contains(permission);
    }

    /**
     * Safe role-name variant: JwtAuthenticationFilter falls back to the literal role "USER"
     * when no backend User record matches the caller (see AuthenticatedPrincipal.role()), which
     * isn't a valid UserRole constant. Treat any unrecognized role name as having no permissions
     * rather than throwing, so an unregistered-but-authenticated caller degrades to "sees nothing
     * gated" instead of a 500.
     */
    public static boolean hasPermission(String roleName, Permission permission) {
        if (roleName == null) {
            return false;
        }
        try {
            return hasPermission(UserRole.valueOf(roleName.toUpperCase()), permission);
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    public static boolean hasPermission(Authentication authentication, Permission permission) {
        if (authentication == null) {
            return false;
        }
        if (!(authentication.getPrincipal() instanceof AuthenticatedPrincipal principal)) {
            return false;
        }
        return hasPermission(principal.role(), permission);
    }
}
